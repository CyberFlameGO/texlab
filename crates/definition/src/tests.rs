use crate::{DefinitionParams, DefinitionResult};

fn check(input: &str) {
    let fixture = test_utils::fixture::Fixture::parse(input);
    let workspace = &fixture.workspace;

    let mut origin_selection_range = None;
    let mut origin_document = None;
    let mut origin_cursor = None;
    for document in &fixture.documents {
        if let Some(cursor) = document.cursor {
            origin_document = Some(document);
            origin_cursor = Some(cursor);
            origin_selection_range = document
                .ranges
                .iter()
                .find(|range| range.contains_inclusive(cursor))
                .copied();

            break;
        }
    }

    let origin_document = origin_document.unwrap();

    let mut expected = Vec::new();
    for document in &fixture.documents {
        let mut ranges = document.ranges.iter();
        while let Some(target_selection_range) = ranges.next().copied() {
            let origin_selection_range = origin_selection_range.unwrap();
            if (&origin_document.uri, origin_selection_range)
                != (&document.uri, target_selection_range)
            {
                expected.push(DefinitionResult {
                    origin_selection_range,
                    target: fixture.workspace.lookup(&document.uri).unwrap(),
                    target_range: *ranges.next().unwrap(),
                    target_selection_range,
                });
            }
        }
    }

    let mut actual = crate::goto_definition(DefinitionParams {
        workspace,
        document: workspace.lookup(&origin_document.uri).unwrap(),
        offset: origin_cursor.unwrap(),
    });

    sort_results(&mut expected);
    sort_results(&mut actual);

    assert_eq!(actual, expected);
}

fn sort_results(items: &mut Vec<DefinitionResult>) {
    items.sort_by(|a, b| {
        let a = (&a.target.uri, a.target_range.start());
        let b = (&b.target.uri, b.target_range.start());
        a.cmp(&b)
    });
}

#[test]
fn test_command_definition() {
    check(
        r#"
%! main.tex
\DeclareMathOperator{\foo}{foo}
                     ^^^^
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
\foo
  |
^^^^"#,
    )
}

#[test]
fn test_document() {
    check(
        r#"
%! foo.tex
\addbibresource{baz.bib}
                  |
                ^^^^^^^

%! bar.bib
@article{foo, bar = {baz}}

%! baz.bib
@article{foo, bar = {baz}}
!
!"#,
    )
}

#[test]
fn test_entry() {
    check(
        r#"
%! foo.tex
\addbibresource{baz.bib}
\cite{foo}
      |
      ^^^

%! bar.bib
@article{foo, bar = {baz}}

%! baz.bib
@article{foo, bar = {baz}}
         ^^^
^^^^^^^^^^^^^^^^^^^^^^^^^^"#,
    )
}

#[test]
fn test_string_simple() {
    check(
        r#"
%! main.bib
@string{foo = {bar}}
        ^^^
^^^^^^^^^^^^^^^^^^^^
@article{bar, author = foo}
                        |
                       ^^^"#,
    )
}

#[test]
fn test_string_join() {
    check(
        r#"
%! main.bib
@string{foo = {bar}}
        ^^^
^^^^^^^^^^^^^^^^^^^^
@article{bar, author = foo # "bar"}
                        |
                       ^^^"#,
    )
}

#[test]
fn test_string_field() {
    check(
        r#"
%! main.bib
@string{foo = {bar}}
@article{bar, author = foo # "bar"}
                |"#,
    )
}
