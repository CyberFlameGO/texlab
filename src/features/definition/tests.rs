use insta::{
    assert_json_snapshot,
    internals::{Content, ContentPath},
};
use lsp_types::Url;

use crate::features::{goto_definition, testing::FeatureTester};

#[test]
fn test_empty_latex_document() {
    let links = goto_definition(
        FeatureTester::builder()
            .files(vec![("main.tex", "")])
            .main("main.tex")
            .line(0)
            .character(0)
            .build()
            .definition(),
    );

    assert_eq!(links, None);
}

#[test]
fn test_empty_bibtex_document() {
    let links = goto_definition(
        FeatureTester::builder()
            .files(vec![("main.bib", "")])
            .main("main.bib")
            .line(0)
            .character(0)
            .build()
            .definition(),
    );

    assert_eq!(links, None);
}

#[test]
fn test_command_definition() {
    let links = goto_definition(
        FeatureTester::builder()
            .files(vec![(
                "main.tex",
                r#"
\DeclareMathOperator{\foo}{foo}
\foo"#,
            )])
            .main("main.tex")
            .line(1)
            .character(2)
            .build()
            .definition(),
    );

    assert_json_snapshot!( links, { "[].targetUri" => insta::dynamic_redaction(redact_uri) });
}

#[test]
fn test_document_simple() {
    let links = goto_definition(
        FeatureTester::builder()
            .files(vec![
                ("foo.tex", r#"\addbibresource{baz.bib}"#),
                ("bar.bib", r#"@article{foo, bar = {baz}}"#),
                ("baz.bib", r#"@article{foo, bar = {baz}}"#),
            ])
            .main("foo.tex")
            .line(0)
            .character(18)
            .build()
            .definition(),
    );

    assert_json_snapshot!( links, { "[].targetUri" => insta::dynamic_redaction(redact_uri) });
}

#[test]
fn test_entry_simple() {
    let links = goto_definition(
        FeatureTester::builder()
            .files(vec![
                (
                    "foo.tex",
                    r#"
\addbibresource{baz.bib}
\cite{foo}"#,
                ),
                ("bar.bib", r#"@article{foo, bar = {baz}}"#),
                ("baz.bib", r#"@article{foo, bar = {baz}}"#),
            ])
            .main("foo.tex")
            .line(1)
            .character(6)
            .build()
            .definition(),
    );

    assert_json_snapshot!( links, { "[].targetUri" => insta::dynamic_redaction(redact_uri) });
}

#[test]
fn test_string_simple() {
    let links = goto_definition(
        FeatureTester::builder()
            .files(vec![(
                "main.bib",
                r#"
@string{foo = {bar}}
@article{bar, author = foo}"#,
            )])
            .main("main.bib")
            .line(1)
            .character(24)
            .build()
            .definition(),
    );

    assert_json_snapshot!( links, { "[].targetUri" => insta::dynamic_redaction(redact_uri) });
}

#[test]
fn test_string_concat() {
    let links = goto_definition(
        FeatureTester::builder()
            .files(vec![(
                "main.bib",
                r#"
@string{foo = {bar}}
@article{bar, author = foo # "bar"}"#,
            )])
            .main("main.bib")
            .line(1)
            .character(24)
            .build()
            .definition(),
    );

    assert_json_snapshot!( links, { "[].targetUri" => insta::dynamic_redaction(redact_uri) });
}

#[test]
fn test_string_field() {
    let links = goto_definition(
        FeatureTester::builder()
            .files(vec![(
                "main.bib",
                r#"
@string{foo = {bar}}
@article{bar, author = foo}"#,
            )])
            .main("main.bib")
            .line(1)
            .character(18)
            .build()
            .definition(),
    );

    assert_eq!(links, None);
}

fn redact_uri(value: Content, _path: ContentPath) -> String {
    value.as_str().unwrap().replace(
        Url::from_directory_path(std::env::temp_dir())
            .unwrap()
            .as_str(),
        "[tmp]/",
    )
}
