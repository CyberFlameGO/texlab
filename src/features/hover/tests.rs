use insta::assert_json_snapshot;

use crate::features::{find_hover, testing::FeatureTester};

#[test]
fn test_empty_latex_document() {
    let hover = find_hover(
        FeatureTester::builder()
            .files(vec![("main.tex", "")])
            .main("main.tex")
            .line(0)
            .character(0)
            .build()
            .hover(),
    );

    assert_eq!(hover, None);
}

#[test]
fn test_empty_bibtex_document() {
    let hover = find_hover(
        FeatureTester::builder()
            .files(vec![("main.bib", "")])
            .main("main.bib")
            .line(0)
            .character(0)
            .build()
            .hover(),
    );

    assert_eq!(hover, None);
}

#[test]
fn test_citation_inside_cite() {
    let hover = find_hover(
        FeatureTester::builder()
            .files(vec![
                (
                    "main.bib",
                    "@article{foo, author = {Foo Bar}, title = {Baz Qux}, year = 1337}",
                ),
                (
                    "main.tex",
                    r#"
\addbibresource{main.bib}
\cite{foo}"#,
                ),
            ])
            .main("main.tex")
            .line(1)
            .character(7)
            .build()
            .hover(),
    );

    assert_json_snapshot!(hover);
}

#[test]
fn test_citation_inside_entry() {
    let hover = find_hover(
        FeatureTester::builder()
            .files(vec![
                (
                    "main.bib",
                    "@article{foo, author = {Foo Bar}, title = {Baz Qux}, year = 1337}",
                ),
                (
                    "main.tex",
                    r#"
\addbibresource{main.bib}
\cite{foo}"#,
                ),
            ])
            .main("main.bib")
            .line(0)
            .character(11)
            .build()
            .hover(),
    );

    assert_json_snapshot!(hover);
}

#[test]
fn test_component_known_package() {
    let hover = find_hover(
        FeatureTester::builder()
            .files(vec![("main.tex", r#"\usepackage{amsmath}"#)])
            .main("main.tex")
            .line(0)
            .character(15)
            .build()
            .hover(),
    );

    assert_json_snapshot!(hover);
}

#[test]
fn test_component_unknown_class() {
    let hover = find_hover(
        FeatureTester::builder()
            .files(vec![("main.tex", r#"\documentclass{abcdefghijklmnop}"#)])
            .main("main.tex")
            .line(0)
            .character(20)
            .build()
            .hover(),
    );

    assert_eq!(hover, None);
}

#[test]
fn test_entry_type_known_type() {
    let hover = find_hover(
        FeatureTester::builder()
            .files(vec![("main.bib", r#"@article{foo,}"#)])
            .main("main.bib")
            .line(0)
            .character(3)
            .build()
            .hover(),
    );

    assert_json_snapshot!(hover);
}

#[test]
fn test_entry_type_unknown_field() {
    let hover = find_hover(
        FeatureTester::builder()
            .files(vec![("main.bib", r#"@foo{bar,}"#)])
            .main("main.bib")
            .line(0)
            .character(3)
            .build()
            .hover(),
    );

    assert_eq!(hover, None);
}

#[test]
fn test_entry_type_key() {
    let hover = find_hover(
        FeatureTester::builder()
            .files(vec![("main.bib", r#"@article{foo,}"#)])
            .main("main.bib")
            .line(0)
            .character(11)
            .build()
            .hover(),
    );

    assert_eq!(hover, None);
}

#[test]
fn test_field_known() {
    let hover = find_hover(
        FeatureTester::builder()
            .files(vec![("main.bib", r#"@article{foo, author = bar}"#)])
            .main("main.bib")
            .line(0)
            .character(15)
            .build()
            .hover(),
    );

    assert_json_snapshot!(hover);
}

#[test]
fn test_field_unknown() {
    let hover = find_hover(
        FeatureTester::builder()
            .files(vec![("main.bib", r#"@article{foo, bar = baz}"#)])
            .main("main.bib")
            .line(0)
            .character(15)
            .build()
            .hover(),
    );

    assert_eq!(hover, None);
}

#[test]
fn test_field_entry_key() {
    let hover = find_hover(
        FeatureTester::builder()
            .files(vec![("main.bib", r#"@article{foo, author = bar}"#)])
            .main("main.bib")
            .line(0)
            .character(11)
            .build()
            .hover(),
    );

    assert_eq!(hover, None);
}

#[test]
fn test_section() {
    let hover = find_hover(
        FeatureTester::builder()
            .files(vec![("main.tex", r#"\section{Foo}\label{sec:foo}"#)])
            .main("main.tex")
            .line(0)
            .character(23)
            .build()
            .hover(),
    );

    assert_json_snapshot!(hover);
}

#[test]
fn test_string_inside_reference() {
    let hover = find_hover(
        FeatureTester::builder()
            .files(vec![(
                "main.bib",
                r#"
@string{foo = "Foo"}
@string{bar = "Bar"}
@article{baz, author = bar}"#,
            )])
            .main("main.bib")
            .line(2)
            .character(24)
            .build()
            .hover(),
    );

    assert_json_snapshot!(hover);
}

#[test]
fn test_string_inside_field() {
    let hover = find_hover(
        FeatureTester::builder()
            .files(vec![(
                "main.bib",
                r#"
@string{foo = "Foo"}
@string{bar = "Bar"}
@article{baz, author = bar}"#,
            )])
            .main("main.bib")
            .line(2)
            .character(21)
            .build()
            .hover(),
    );

    assert_eq!(hover, None);
}
