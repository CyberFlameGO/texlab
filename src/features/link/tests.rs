use insta::assert_json_snapshot;

use crate::features::{find_document_links, redactions::redact_uri, testing::FeatureTester};

#[test]
fn test_empty_latex_document() {
    let links = find_document_links(
        FeatureTester::builder()
            .files(vec![("main.tex", "")])
            .main("main.tex")
            .build()
            .link(),
    );

    assert_eq!(links, Vec::new());
}

#[test]
fn test_empty_bibtex_document() {
    let links = find_document_links(
        FeatureTester::builder()
            .files(vec![("main.bib", "")])
            .main("main.bib")
            .build()
            .link(),
    );

    assert_eq!(links, Vec::new());
}

#[test]
fn test_includes() {
    let links = find_document_links(
        FeatureTester::builder()
            .files(vec![("foo.tex", r#"\input{bar.tex}"#), ("bar.tex", r#""#)])
            .main("foo.tex")
            .build()
            .link(),
    );

    assert_json_snapshot!(
        links,
        {
            "[].target" => insta::dynamic_redaction(redact_uri),
            "." => insta::sorted_redaction(),
        }
    );
}

#[test]
fn test_imports() {
    let links = find_document_links(
        FeatureTester::builder()
            .files(vec![
                ("foo.tex", r#"\import{bar/}{baz}"#),
                ("bar/baz.tex", r#""#),
            ])
            .main("foo.tex")
            .build()
            .link(),
    );

    assert_json_snapshot!(
        links,
        {
            "[].target" => insta::dynamic_redaction(redact_uri),
            "." => insta::sorted_redaction(),
        }
    );
}
