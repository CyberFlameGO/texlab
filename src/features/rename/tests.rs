use insta::assert_json_snapshot;

use crate::features::{redactions::redact_uri, testing::FeatureTester};

use super::rename_all;

#[test]
fn test_command() {
    let edit = rename_all(
        FeatureTester::builder()
            .files(vec![
                ("foo.tex", r#"\baz\include{bar.tex}"#),
                ("bar.tex", r#"\baz"#),
            ])
            .main("foo.tex")
            .line(0)
            .character(2)
            .new_name("qux")
            .build()
            .rename(),
    );

    assert_json_snapshot!(
        edit,
        {
            ".changes.$key" => insta::dynamic_redaction(redact_uri),
            ".changes" => insta::sorted_redaction(),
        }
    );
}

#[test]
fn test_entry() {
    let edit = rename_all(
        FeatureTester::builder()
            .files(vec![
                ("main.bib", r#"@article{foo, bar = baz}"#),
                ("main.tex", "\\addbibresource{main.bib}\n\\cite{foo}"),
            ])
            .main("main.bib")
            .line(0)
            .character(9)
            .new_name("qux")
            .build()
            .rename(),
    );

    assert_json_snapshot!(
        edit,
        {
            ".changes.$key" => insta::dynamic_redaction(redact_uri),
            ".changes" => insta::sorted_redaction(),
        }
    );
}

#[test]
fn test_citation() {
    let edit = rename_all(
        FeatureTester::builder()
            .files(vec![
                ("main.bib", r#"@article{foo, bar = baz}"#),
                ("main.tex", "\\addbibresource{main.bib}\n\\cite{foo}"),
            ])
            .main("main.tex")
            .line(1)
            .character(6)
            .new_name("qux")
            .build()
            .rename(),
    );

    assert_json_snapshot!(
        edit,
        {
            ".changes.$key" => insta::dynamic_redaction(redact_uri),
            ".changes" => insta::sorted_redaction(),
        }
    );
}

#[test]
fn test_label() {
    let edit = rename_all(
        FeatureTester::builder()
            .files(vec![
                ("foo.tex", r#"\label{foo}\include{bar}"#),
                ("bar.tex", r#"\ref{foo}"#),
                ("baz.tex", r#"\ref{foo}"#),
            ])
            .main("foo.tex")
            .line(0)
            .character(7)
            .new_name("bar")
            .build()
            .rename(),
    );

    assert_json_snapshot!(
        edit,
        {
            ".changes.$key" => insta::dynamic_redaction(redact_uri),
            ".changes" => insta::sorted_redaction(),
        }
    );
}
