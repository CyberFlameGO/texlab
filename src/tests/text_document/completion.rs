use insta::assert_json_snapshot;
use lsp_types::{
    request::{Completion, ResolveCompletionItem},
    CompletionItem, CompletionList, CompletionParams, CompletionResponse, CompletionTextEdit,
    Range,
};

use crate::tests::{client::Client, fixture};

fn complete(fixture: &str) -> Vec<CompletionItem> {
    let mut client = Client::spawn();
    client.initialize(
        serde_json::from_value(serde_json::json!({
            "textDocument": {
                "completion": {
                    "completionItem": {
                        "documentationFormat": ["plaintext", "markdown"]
                    }
                }
            }
        }))
        .unwrap(),
        None,
    );

    let fixture = fixture::parse(fixture);
    for file in fixture.files {
        client.open(file.name, file.lang, file.text);
    }

    let range = fixture
        .ranges
        .values()
        .next()
        .and_then(|map| map.values().next())
        .map_or_else(
            || {
                let position = fixture.cursor.unwrap().position;
                Range::new(position, position)
            },
            |file_range| file_range.range,
        );

    let actual_list = client
        .request::<Completion>(CompletionParams {
            text_document_position: fixture.cursor.unwrap().into_params(&client),
            partial_result_params: Default::default(),
            work_done_progress_params: Default::default(),
            context: None,
        })
        .unwrap()
        .map_or(CompletionList::default(), |actual| match actual {
            CompletionResponse::List(list) => list,
            CompletionResponse::Array(_) => unreachable!(),
        });

    for item in &actual_list.items {
        if let Some(CompletionTextEdit::Edit(edit)) = item.text_edit.as_ref() {
            assert_eq!(edit.range, range);
        }
    }

    let actual_items = actual_list
        .items
        .into_iter()
        .take(5)
        .map(|item| client.request::<ResolveCompletionItem>(item).unwrap())
        .collect();

    client.shutdown();
    actual_items
}

macro_rules! assert_items {
    ($items:expr) => {
        assert_json_snapshot!($items, {
            "[].data" => "[data]",
            "[].sortText" => "[sortText]",
            "[].documentation" => "[documentation]",
            "[].textEdit.range" => "[range]",
            "[]" => insta::sorted_redaction()
        });
    };
}

#[test]
fn acronym_ref_simple() {
    assert_items!(complete(
        r#"
%TEX main.tex
%SRC \newacronym[longplural={Frames per Second}]{fpsLabel}{FPS}{Frame per Second}
%SRC \acrshort{f}
%CUR           ^
%1.1           ^"#
    ));
}

#[test]
fn acronym_ref_empty() {
    assert_items!(complete(
        r#"
%TEX main.tex
%SRC \newacronym[longplural={Frames per Second}]{fpsLabel}{FPS}{Frame per Second}
%SRC \acrshort{}
%CUR           ^"#
    ));
}

#[test]
fn acronym_ref_after_group() {
    let actual_items = complete(
        r#"
%TEX main.tex
%SRC \newacronym[longplural={Frames per Second}]{fpsLabel}{FPS}{Frame per Second}
%SRC \acrshort{}
%CUR            ^
%1.1            ^"#,
    );

    assert_eq!(actual_items, Vec::new());
}

#[test]
fn acronym_ref_open_brace() {
    assert_items!(complete(
        r#"
%TEX main.tex
%SRC \newacronym[longplural={Frames per Second}]{fpsLabel}{FPS}{Frame per Second}
%SRC \acrshort{f
%CUR           ^
%1.1           ^"#
    ));
}

#[test]
fn acronym_package_ref() {
    assert_items!(complete(
        r#"
%TEX main.tex
%SRC \acrodef{fpsLabel}[FPS]{Frames per Second}
%SRC \ac{f
%CUR     ^
%1.1     ^"#
    ));
}

#[test]
fn glossary_ref_simple() {
    assert_items!(complete(
        r#"
%TEX main.tex
%SRC \newacronym[longplural={Frames per Second}]{fpsLabel}{FPS}{Frame per Second}
%SRC \gls{f}
%CUR      ^
%1.1      ^"#
    ));
}

#[test]
fn glossary_ref_open_brace() {
    assert_items!(complete(
        r#"
%TEX main.tex
%SRC \newacronym[longplural={Frames per Second}]{fpsLabel}{FPS}{Frame per Second}
%SRC \gls{f
%CUR      ^
%1.1      ^"#
    ));
}

#[test]
fn argument_empty() {
    assert_items!(complete(
        r#"
%TEX main.tex
%SRC \usepackage{amsfonts}
%SRC \mathbb{}
%CUR         ^"#
    ));
}

#[test]
fn argument_word() {
    assert_items!(complete(
        r#"
%TEX main.tex
%SRC \usepackage{amsfonts}
%SRC \mathbb{A}
%CUR         ^
%1.1         ^"#
    ));
}

#[test]
fn argument_open_brace() {
    assert_items!(complete(
        r#"
%TEX main.tex
%SRC \usepackage{amsfonts}
%SRC \mathbb{
%CUR         ^
%SRC Test"#
    ));
}

#[test]
fn argument_open_brace_unrelated() {
    let actual_items = complete(
        r#"
%TEX main.tex
%SRC \usepackage{amsfonts}
%SRC \mathbb{}{
%CUR           ^
%SRC Test"#,
    );

    assert_eq!(actual_items, Vec::new());
}

#[test]
fn begin_environment_without_snippet_support() {
    assert_items!(complete(
        r#"
%TEX main.tex
%SRC \beg
%CUR     ^
%1.1  ^^^"#
    ));
}

#[test]
fn citation() {
    assert_items!(complete(
        r#"
%TEX main.tex
%SRC \documentclass{article}
%SRC \bibliography{main}
%SRC \begin{document}
%SRC \cite{
%CUR       ^
%SRC \end{document}

%BIB main.bib
%SRC @article{foo:2019,
%SRC     author = {Foo Bar},
%SRC     title = {Baz Qux},
%SRC     year = {2019},
%SRC }
%SRC 
%SRC @article{bar:2005,}"#
    ));
}

#[test]
fn citation_open_brace() {
    assert_items!(complete(
        r#"
%TEX main.tex
%SRC \addbibresource{main.bib}
%SRC \cite{
%CUR       ^

%BIB main.bib
%SRC @article{foo,}"#
    ));
}

#[test]
fn citation_open_brace_multiple() {
    assert_items!(complete(
        r#"
%TEX main.tex
%SRC \addbibresource{main.bib}
%SRC \cite{foo,a
%CUR           ^
%1.1           ^

%BIB main.bib
%SRC @article{foo,}"#
    ));
}

#[test]
fn citation_acronym() {
    assert_items!(complete(
        r#"
%TEX main.tex
%SRC \addbibresource{main.bib}
%SRC \DeclareAcronym{foo}{cite={}}
%CUR                            ^

%BIB main.bib
%SRC @article{foo,}"#
    ));
}

#[test]
fn citation_after_brace() {
    let actual_items = complete(
        r#"
%TEX main.tex
%SRC \documentclass{article}
%SRC \bibliography{main}
%SRC \begin{document}
%SRC \cite{}
%CUR        ^
%SRC \end{document}

%BIB main.bib
%SRC @article{foo,}"#,
    );

    assert_eq!(actual_items, Vec::new());
}

#[test]
fn color_model_definition_simple() {
    assert_items!(complete(
        r#"
%TEX main.tex
%SRC \definecolor{foo}{}
%CUR                   ^"#
    ));
}

#[test]
fn color_model_definition_open_brace() {
    assert_items!(complete(
        r#"
%TEX main.tex
%SRC \definecolor{foo}{
%CUR                   ^"#
    ));
}

#[test]
fn color_model_definition_set_simple() {
    assert_items!(complete(
        r#"
%TEX main.tex
%SRC \definecolorset{}
%CUR                 ^"#
    ));
}

#[test]
fn color_model_definition_set_open_brace() {
    assert_items!(complete(
        r#"
%TEX main.tex
%SRC \definecolorset{
%CUR                 ^"#
    ));
}

#[test]
fn color_simple() {
    assert_items!(complete(
        r#"
%TEX main.tex
%SRC \color{}
%CUR        ^"#
    ));
}

#[test]
fn color_word() {
    assert_items!(complete(
        r#"
%TEX main.tex
%SRC \color{re}
%CUR         ^
%1.1        ^^"#
    ));
}

#[test]
fn color_open_brace() {
    assert_items!(complete(
        r#"
%TEX main.tex
%SRC \color{
%CUR        ^"#
    ));
}

#[test]
fn component_command_simple() {
    assert_items!(complete(
        r#"
%TEX main.tex
%SRC \
%CUR  ^"#
    ));
}

#[test]
fn component_command_simple_before() {
    let actual_items = complete(
        r#"
%TEX main.tex
%SRC \
%CUR ^"#,
    );

    assert_eq!(actual_items, Vec::new());
}

#[test]
fn component_command_simple_package() {
    assert_items!(complete(
        r#"
%TEX main.tex
%SRC \usepackage{lipsum}
%SRC \lips
%CUR    ^
%1.1  ^^^^"#
    ));
}

#[test]
fn component_command_bibtex() {
    assert_items!(complete(
        r#"
%BIB main.bib
%SRC @article{b, 
%SRC     c = {\LaT }
%CUR            ^
%1.1           ^^^
%SRC }"#
    ));
}

#[test]
fn component_environment_simple() {
    assert_items!(complete(
        r#"
%TEX main.tex
%SRC \begin{doc
%CUR           ^
%1.1        ^^^"#
    ));
}

#[test]
fn component_environment_simple_end() {
    assert_items!(complete(
        r#"
%TEX main.tex
%SRC \begin{document}
%SRC \end{
%CUR      ^"#
    ));
}

#[test]
fn component_environment_class() {
    assert_items!(complete(
        r#"
%TEX main.tex
%SRC \documentclass{article}
%SRC \begin{thein}
%CUR           ^
%1.1        ^^^^^"#
    ));
}

#[test]
fn component_environment_command_definition() {
    assert_items!(complete(
        r#"
%TEX main.tex
%SRC \newcommand{\foo}{\begin{doc}
%CUR                            ^
%1.1                          ^^^"#
    ));
}

#[test]
fn entry_type_at_empty() {
    assert_items!(complete(
        r#"
%BIB main.bib
%SRC @
%CUR  ^"#
    ));
}

#[test]
fn entry_type_before_preamble() {
    assert_items!(complete(
        r#"
%BIB main.bib
%SRC @preamble
%CUR  ^
%1.1  ^^^^^^^^"#
    ));
}

#[test]
fn entry_type_before_string() {
    assert_items!(complete(
        r#"
%BIB main.bib
%SRC @string
%CUR  ^
%1.1  ^^^^^^"#
    ));
}

#[test]
fn entry_type_before_article() {
    assert_items!(complete(
        r#"
%BIB main.bib
%SRC @article
%CUR  ^
%1.1  ^^^^^^^"#
    ));
}

#[test]
fn entry_type_after_preamble() {
    assert_items!(complete(
        r#"
%BIB main.bib
%SRC @preamble{
%CUR          ^
%1.1  ^^^^^^^^"#
    ));
}

#[test]
fn entry_type_after_string() {
    assert_items!(complete(
        r#"
%BIB main.bib
%SRC @string{
%CUR        ^
%1.1  ^^^^^^"#
    ));
}

#[test]
fn entry_type_complete_entry() {
    assert_items!(complete(
        r#"
%BIB main.bib
%SRC @article{foo, author = {foo}}
%CUR    ^
%1.1  ^^^^^^^"#
    ));
}

#[test]
fn field_empty_entry_open() {
    assert_items!(complete(
        r#"
%BIB main.bib
%SRC @article{foo,
%CUR              ^"#
    ));
}

#[test]
fn field_empty_entry_closed() {
    assert_items!(complete(
        r#"
%BIB main.bib
%SRC @article{foo,}
%CUR              ^"#
    ));
}

#[test]
fn field_entry_field_name() {
    assert_items!(complete(
        r#"
%BIB main.bib
%SRC @article{foo, a
%CUR                ^
%1.1               ^"#
    ));
}

#[test]
fn field_entry_two_fields_name_open() {
    assert_items!(complete(
        r#"
%BIB main.bib
%SRC @article{foo, author = bar, edit
%CUR                              ^
%1.1                             ^^^^"#
    ));
}

#[test]
fn field_entry_two_fields_name_closed() {
    assert_items!(complete(
        r#"
%BIB main.bib
%SRC @article{foo, author = bar, edit}
%CUR                              ^
%1.1                             ^^^^"#
    ));
}

#[test]
fn import_package_open_brace() {
    assert_items!(complete(
        r#"
%TEX main.tex
%SRC \usepackage{lips
%CUR              ^
%1.1             ^^^^"#
    ));
}

#[test]
fn import_package_closed_brace() {
    assert_items!(complete(
        r#"
%TEX main.tex
%SRC \usepackage{lips}
%CUR              ^
%1.1             ^^^^"#
    ));
}

#[test]
fn import_class_open_brace() {
    assert_items!(complete(
        r#"
%TEX main.tex
%SRC \documentclass{art \foo
%CUR                 ^
%1.1                ^^^"#
    ));
}

#[test]
fn import_class_closed_brace() {
    assert_items!(complete(
        r#"
%TEX main.tex
%SRC \documentclass{art}
%CUR                 ^
%1.1                ^^^"#
    ));
}

#[test]
fn label() {
    assert_items!(complete(
        r#"
%TEX foo.tex
%SRC \documentclass{article}
%SRC 
%SRC \usepackage{amsmath}
%SRC \usepackage{caption}
%SRC \usepackage{amsthm}
%SRC \newtheorem{lemma}{Lemma}
%SRC 
%SRC \begin{document}
%SRC 
%SRC \section{Foo}%
%SRC \label{sec:foo}
%SRC 
%SRC \begin{equation}%
%SRC \label{eq:foo}
%SRC     1 + 1 = 2
%SRC \end{equation}
%SRC 
%SRC \begin{equation}%
%SRC \label{eq:bar}
%SRC     1 + 1 = 2
%SRC \end{equation}
%SRC 
%SRC \begin{figure}%
%SRC \LaTeX{}
%SRC \caption{Baz}%
%SRC \label{fig:baz}
%SRC \end{figure}
%SRC 
%SRC \begin{lemma}%
%SRC \label{thm:foo}
%SRC     1 + 1 = 2
%SRC \end{lemma}
%SRC 
%SRC \include{bar}
%SRC 
%SRC \end{document}

%TEX bar.tex
%SRC \section{Bar}%
%SRC \label{sec:bar}
%SRC 
%SRC Lorem ipsum dolor sit amet.
%SRC \ref{}
%CUR      ^

%TEX foo.aux
%SRC \relax
%SRC \@writefile{lof}{\contentsline {figure}{\numberline {1}{\ignorespaces Baz\relax }}{1}\protected@file@percent }
%SRC \providecommand*\caption@xref[2]{\@setref\relax\@undefined{#1}}
%SRC \newlabel{fig:baz}{{1}{1}}
%SRC \@writefile{toc}{\contentsline {section}{\numberline {1}Foo}{1}\protected@file@percent }
%SRC \newlabel{sec:foo}{{1}{1}}
%SRC \newlabel{eq:foo}{{1}{1}}
%SRC \newlabel{eq:bar}{{2}{1}}
%SRC \newlabel{thm:foo}{{1}{1}}
%SRC \@input{bar.aux}"#
    ));
}

#[test]
fn theorem_begin() {
    assert_items!(complete(
        r#"
%TEX main.tex
%SRC \newtheorem{lemma}{Lemma}
%SRC \begin{lem 
%CUR         ^
%1.1        ^^^"#
    ));
}

#[test]
fn theorem_end() {
    assert_items!(complete(
        r#"
%TEX main.tex
%SRC \newtheorem{lemma}{Lemma}
%SRC \begin{}
%SRC \end{lem
%CUR       ^
%1.1      ^^^"#
    ));
}

#[test]
fn tikz_library_open_brace() {
    assert_items!(complete(
        r#"
%TEX main.tex
%SRC \usepgflibrary{
%CUR                ^"#
    ));
}

#[test]
fn tikz_library_closed_brace() {
    assert_items!(complete(
        r#"
%TEX main.tex
%SRC \usepgflibrary{}
%CUR                ^"#
    ));
}

#[test]
fn test_user_command() {
    assert_items!(complete(
        r#"
%TEX main.tex
%SRC \foobar
%SRC \fooba
%CUR    ^
%1.1  ^^^^^
%SRC \begin{foo}
%SRC \end{foo}
%SRC \begin{fo}
"#
    ));
}

#[test]
fn test_user_environment() {
    assert_items!(complete(
        r#"
%TEX main.tex
%SRC \foobar
%SRC \fooba
%SRC \begin{foo}
%SRC \end{foo}
%SRC \begin{fo}
%CUR         ^
%1.1        ^^
"#
    ));
}
