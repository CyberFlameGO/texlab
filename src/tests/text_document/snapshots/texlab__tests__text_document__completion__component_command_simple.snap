---
source: src/tests/text_document/completion.rs
expression: "complete(r#\"\n%TEX main.tex\n%SRC \\\n%CUR  ^\"#)?"
---
[
  {
    "detail": "built-in",
    "kind": 1,
    "label": "!",
    "preselect": false,
    "sortText": "[sortText]",
    "textEdit": {
      "range": "[range]",
      "newText": "!"
    }
  },
  {
    "detail": "built-in",
    "documentation": "[documentation]",
    "kind": 1,
    "label": "\"",
    "preselect": false,
    "sortText": "[sortText]",
    "textEdit": {
      "range": "[range]",
      "newText": "\""
    }
  },
  {
    "detail": "built-in",
    "kind": 1,
    "label": "#",
    "preselect": false,
    "sortText": "[sortText]",
    "textEdit": {
      "range": "[range]",
      "newText": "#"
    }
  },
  {
    "detail": "$, built-in",
    "documentation": "[documentation]",
    "kind": 1,
    "label": "$",
    "preselect": false,
    "sortText": "[sortText]",
    "textEdit": {
      "range": "[range]",
      "newText": "$"
    }
  },
  {
    "detail": "built-in",
    "kind": 1,
    "label": "%",
    "preselect": false,
    "sortText": "[sortText]",
    "textEdit": {
      "range": "[range]",
      "newText": "%"
    }
  }
]
